Update MovingAvg function

rraustad opened the pull request:
Pull request overview
---------------------
 - Updates MovingAvg function to improve speed as part of #9583
 - The MovingAvg function 0's the output array internally (which eliminates the need to zero output array prior to call), creates a temporary array 60% larger than needed and requires the output array to be equated to the original array after the call.

    state.dataZoneEquipmentManager->AvgData = 0.0;
    MovingAvg(state.dataSize->CalcZoneSizing(state.dataSize->CurOverallSimDay, CtrlZoneNum).CoolFlowSeq,
              state.dataZoneEquipmentManager->NumOfTimeStepInDay,
              state.dataSize->NumTimeStepsInAvg,
              state.dataZoneEquipmentManager->AvgData);
    state.dataSize->CalcZoneSizing(state.dataSize->CurOverallSimDay, CtrlZoneNum).CoolFlowSeq = state.dataZoneEquipmentManager->AvgData;

 - A new function was created to reduce the temporary array size and act directly on the original array.

    void MovingAvg2(Array1D<Real64> &DataIn,      // input data that needs smoothing
                   int const NumItemsInAvg              // number of items in the averaging window
    );

    MovingAvg2(state.dataSize->CalcZoneSizing(state.dataSize->CurOverallSimDay, CtrlZoneNum).CoolFlowSeq,
                         state.dataSize->NumTimeStepsInAvg);

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

rraustad committed:
Update MovingAvg function

rraustad commented on src/EnergyPlus/General.cc:
```diff
@@ -398,6 +398,45 @@ void MovingAvg(Array1A<Real64> const DataIn, // input data that needs smoothing
+    // Note that DataIn and SmoothedData should have the same size. This is the reponsibility
+    // of the calling routine. NumItemsInAvg should be no bigger than the size of DataIn.
+
+    if (NumItemsInAvg == 1) return;
```
Data does not need to be smoothed.

rraustad commented on src/EnergyPlus/General.cc:
```diff
@@ -398,6 +398,45 @@ void MovingAvg(Array1A<Real64> const DataIn, // input data that needs smoothing
+    }
+    for (int i = 1; i <= NumItemsInAvg; ++i) {
+        TempData(NumItemsInAvg - i + 1) = DataIn(DataIn.size() - i + 1);
+    }
```
Not sure how to do this more directly.

rraustad commented on src/EnergyPlus/General.cc:
```diff
@@ -398,6 +398,45 @@ void MovingAvg(Array1A<Real64> const DataIn, // input data that needs smoothing
+    //       AUTHOR         Fred Buhl
+    //       DATE WRITTEN   January 2003
+    //       MODIFIED       na
+    //       RE-ENGINEERED  na
```
Clean up comments after discussion of function methods.

rraustad commented on src/EnergyPlus/General.cc:
```diff
@@ -398,6 +398,45 @@ void MovingAvg(Array1A<Real64> const DataIn, // input data that needs smoothing
+        }
+        DataIn(i) /= double(NumItemsInAvg);
+    }
+}
```
Acts directly on original array. Starting at end of bottom temp array (working up through top temp array), add element i-1 to i, and i-2 to i, etc., then averages those data to element i.

rraustad commented:
This new function is almost twice as fast as original code. 13 seconds vs 8 seconds for 100,000,000 calls. This includes removing the existing reset of AvgData= 0.0 before call and setting OriginalData = AvgData after call.

![image](https://user-images.githubusercontent.com/5778156/184456291-036f701a-5d9a-4933-9f18-dbaa5a7f7116.png)

    std::chrono::system_clock::time_point startMovingAvg = std::chrono::system_clock::now();
    for (int loop = 1; loop < 100000000; ++loop) {
        outputData = 0;
        MovingAvg(saveData, numItem, avgWindowWidth, outputData);
        inputData = outputData;
    }
    std::chrono::system_clock::time_point endMovingAvg = std::chrono::system_clock::now();
    for (int loop = 1; loop < 100000000; ++loop) {
        MovingAvg2(saveData, avgWindowWidth);
    }
    std::chrono::system_clock::time_point endMovingAvg2 = std::chrono::system_clock::now();

    time_t startt = std::chrono::system_clock::to_time_t(startMovingAvg);
    time_t endt = std::chrono::system_clock::to_time_t(endMovingAvg);
    time_t endt2 = std::chrono::system_clock::to_time_t(endMovingAvg2);

    printf((ctime(&startt)));
    printf((ctime(&endt)));
    printf((ctime(&endt2)));

rraustad commented:
Waiting on #9584 to finish up this transition to the improved MovingAvg function.

rraustad committed:
Clean up

rraustad committed:
Streamline

rraustad committed:
Optimize loop counter

rraustad committed:
Update another moving avg function

rraustad committed:
Update MovingAvg calls

rraustad committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into Update-MovingAvg

Myoldmopar commented:
@rraustad are you expecting diffs from this refactor?  I pulled develop into this branch locally to resolve diffs and got a small amount of big math diffs (less than what is showing on the last CI run, but actually more small math diffs).

rraustad commented:
I was not expecting diffs but after I pulled in develop and fixed the conflict, diffs showed up. I was checking diffs along the way but I guess I did not check after converting ZoneEquipmentManager. I am not sure if this has something to do with the original Array1A usage and now Array1D, or something else. I reverted ZoneEquipmentManager and there were no diffs. After converting the first call to MovingAvg diffs showed up right off the bat. I single stepped the first call and found this. adataA is the value after the averaging, adataB is value calculated using ((element(i) + element(i-1) + element(i-2) + element(i-3))/4. adatai is the element index. In develop element 79 shows a difference before and after while this branch does not. I took develop and just changed MovingAvg arrays to Array1D and get diffs. So I think I converted the function correctly, there is just something going on that I don't understand yet.

![image](https://user-images.githubusercontent.com/5778156/184931527-1862b654-bf32-4f33-bdff-b71684330442.png)

Note how the data looks in the debugger. element 76, which is used in the average of element 79 (avg window = 4), has a different number of decimals. Not sure if that has anything to do with the price of tea but it is sure coincidental. Also note that after element 79, the averaged data appears equal to the calculated value. So I'm not sure what to do at this point. Also, the diffs are such that I would need to figure this out as opposed to saying "most of the diffs are small" and merging this. It might be one of those fix develop first and then check this branch for diffs.

![image](https://user-images.githubusercontent.com/5778156/184932492-312752e3-da89-4b45-b591-9b87c2a9dfb3.png)

rraustad commented:
Also, if the performance indicator is true (-0.2) then there is no need for this change. I'm not sure I believe that with diffs involved.

rraustad commented:
@Myoldmopar I started over and now it looks like there are no diffs (~20 files). The MovingAvg function is not as tight as it was before but still better. Lot's of other clean up here as well. I'll run local regressions first before pushing this last change.

rraustad committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into Update-MovingAvg

rraustad committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into Update-MovingAvg

rraustad commented:
Windows runs fine here, that CI has a problem. Unsigned/signed warning 2 places from `for (int i = 1; i <= DataIn.size(); ++i) {` but more disturbing is performance of -0.01%. I don't understand that. The compiler must be optimizing out most of these changes.

![image](https://user-images.githubusercontent.com/5778156/185104053-59062ad6-7091-41e7-aa8f-dd5ca6e65668.png)

rraustad committed:
Fix warnings. Revert attempt to optimize.

rraustad committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into Update-MovingAvg

rraustad commented:
I am not smarter than a 5th grader, or the compiler. https://quick-bench.com/q/Ieghg1I1RR8kfJkGqYRiXBitDDo

rraustad committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into Update-MovingAvg

rraustad commented on src/EnergyPlus/General.cc:
```diff
@@ -421,44 +421,22 @@ std::string &strip_trailing_zeros(std::string &InputString)
-               int const NumItemsInAvg,      // number of items in the averaging window
-               Array1A<Real64> SmoothedData  // output data after smoothing
-)
+void MovingAvg(Array1D<Real64> &DataIn, int const NumItemsInAvg)
```
This function was relying on a temp array to calculate the smooted array. The change acts directly on the referenced array and gets rid of 2 arguments.

rraustad commented on src/EnergyPlus/General.cc:
```diff
@@ -421,44 +421,22 @@ std::string &strip_trailing_zeros(std::string &InputString)
-)
+void MovingAvg(Array1D<Real64> &DataIn, int const NumItemsInAvg)
 {
+    if (NumItemsInAvg <= 1) return; // no need to average/smooth
```
Smoothing does not happen if averaging window  = 1

rraustad commented on src/EnergyPlus/General.cc:
```diff
@@ -421,44 +421,22 @@ std::string &strip_trailing_zeros(std::string &InputString)
+            DataIn(i) += TempData(DataIn.size() - NumItemsInAvg + i + j); // sum top end including NumItemsInAvg history terms
         }
-        SmoothedData(i) /= double(NumItemsInAvg);
+        DataIn(i) /= NumItemsInAvg; // average to smooth over NumItemsInAvg window
```
Does the same thing as before except does not size the temp array as 3x the original. Uses 2x, which is still too large, but works fine just the same.

rraustad commented on src/EnergyPlus/OutputReportTabular.cc:
```diff
@@ -16046,96 +16035,100 @@ void ComputeTableBodyUsingMovingAvg(EnergyPlusData &state,
-            MovingAvgAtMaxTime(state, ort->peopleInstantSeq(desDaySelected, _, zoneIndex), NumOfTimeStepInDay, timeOfMax);
+        AvgData = state.dataOutRptTab->peopleInstantSeq(desDaySelected, _, zoneIndex);
+        General::MovingAvg(AvgData, state.dataSize->NumTimeStepsInAvg);
+        resultCells(LoadCompCol::SensInst, LoadCompRow::People) = AvgData(timeOfMax);
```
This area avoids acting on the original array, which adds a few lines of code. At this point in the code, could have probably acted directly on the original array but to avoid diffs this method was chosen. This does however do the same thing as done before.

mitchute commented:
This looks OK. I ran regressions and unit tests locally and it looks good. Merging.
